marker_api/src/ast/stmt.rs

```diff
@@ -1,24 +1,37 @@
-use crate::{context::with_cx, ffi::FfiOption, CtorBlocker};
+use std::fmt::Debug;
+use std::marker::PhantomData;
 
-use super::{
-    expr::ExprKind, item::ItemKind, pat::PatKind, ty::SynTyKind, LetStmtId, Span, SpanId, StmtId, StmtIdInner,
-};
+use crate::{ffi::FfiOption, private::Sealed};
+
+use super::{expr::ExprKind, item::ItemKind, pat::PatKind, ty::SynTyKind, Span, SpanId, StmtId};
+
+/// This trait combines methods, which all statements have in common.
+///
+/// This trait is only meant to be implemented inside this crate. The `Sealed`
+/// super trait prevents external implementations.
+pub trait StmtData<'ast>: Debug + Sealed {
+    /// Returns the [`SpanId`] of this statement
+    fn id(&self) -> StmtId;
+
+    /// Returns the [`Span`] of this statement.
+    fn span(&self) -> &Span<'ast>;
+}
 
 #[repr(C)]
 #[non_exhaustive]
 #[derive(Debug, Copy, Clone)]
 pub enum StmtKind<'ast> {
-    Item(&'ast ItemKind<'ast>, CtorBlocker),
+    Item(&'ast ItemStmt<'ast>),
```

Comment:
Did you consider following the naming used in `syn` where you can see a pattern [like this](https://docs.rs/syn/2.0.33/syn/enum.Expr.html#variants):
```rust
enum Expr {
    Array(ExprArray),
    Assign(ExprAssign),
    // ...
}
```
However, a bit different naming should be used if some of the types may be reused in different contexts. For example, items not always have to appear inside of statements, os they shouldn't be called `ItemStmt` or `StmtItem`, but just `Item`. `Let` bindings may occur as part of `if/while let` , so maybe that part should also be just `Let`.
```rust
enum Stmt {
    Let(Let),
    Item(Item),
}
```
I'm definitely is wrong in this thinking because "let statement" also includes a semicolon at the end...  I haven't yet fully grasped the naming patterns in `marker_api`, so correct me if I'm wrong and this code uses an already-established extensible and workable naming scheme

Reply:
Interesting that `syn` uses a similar AST structure, which I would have found that early on. :)

Marker's AST structure usually has a `*Kind` enum at the top level. This was inspired from rustc, which also uses `*Kind` enum, in nodes (See [`Expr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.Expr.html) with the `.kind` fiel of type [`ExprKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.ExprKind.html)). It was called `Expr` for some time, but that felt a bit off to me for some reason. Guess it felt like the enums is not the actual expression, since it doesn't hold any data and just the different variant structs do.

The variant structs always have the type at the end. So any enum wrapped by `ExprKind` should end with `Expr`. This means an `if` expression is called `IfExpr`. Having the type in the end makes it more natural to read IMO. If you compare these two:

```
// Reads to me as "a takes a parameter `x` with the type `ExprIf`"
fn a(x: &ExprIf<'ast>) {}

// Reads to me as "a takes an if expression as parameter `x`"
fn a(x: &IfExpr<'ast>) {}
```

When I started the AST representation, I tried to find some examples, but didn't find any (And apparently ignored `syn` at that time). So this naming convention is just a result of what felt right and what I'm used to from rustc and [javaparser](https://javadoc.io/doc/com.github.javaparser/javaparser-core/latest/index.html) which I also used in a different project beforehand.